lib/cracker.rb

```diff
@@ -1,3 +1,61 @@
-class Cracker
+require './lib/decoder'
 
+class Cracker < Decoder
+
+  def reverse_shifts(ciphertext, known_end = " end")
+    last_four = ciphertext[-known_end.length..-1].split('')
+    last_four.zip(known_end.split('')).map do |cipher_char, known_char|
+      shift = @whitelist.index(cipher_char) - @whitelist.index(known_char)
+      shift.negative? ? (shift + @whitelist.length) : shift
+    end.rotate(-(ciphertext.length % known_end.length))
+  end
+
+  def root_codes(shifts, offsets)
+    shifts.zip(offsets).map {|shift, offset| shift - offset.to_i}
+  end
+
+  def full_shift(code)
+    code + @whitelist.length
+  end
+
+  def code_to_string(code)
+    code.to_s.length < 2 ? code.to_s.prepend("0") : code.to_s
+  end
+
+  def potential_codes(root_codes)
+    root_codes.map do |code|
+      codes = [code]
+      codes << full_shift(codes.last) until full_shift(codes.last) > 99
+      codes.map {|potential_code| code_to_string(potential_code)}
+    end
+  end
+
+  def all_combinations(code_arrays)
+    code_arrays[0].product(code_arrays[1], code_arrays[2], code_arrays[3])
+  end
+
+  def chains_together?(combination)
+    combination.each_cons(2).all? {|code1, code2| code1[1] == code2[0]}
+  end
+
+  def valid_codes(all_combinations)
+    all_combinations.find_all {|combination| chains_together?(combination)}
+  end
+
+  def key_from_codes(codes)
+    codes.reduce("") {|acc, key| acc << key[0]} << codes.last[-1]
```

Comment:
makes a key out of codes by adding the first digit of each code as well as the entire final code. also 
```suggestion
    codes.reduce("") {|acc, code| acc << code[0]} << codes.last[-1]
```